KLR/NKI/Types.lean

```diff
@@ -0,0 +1,54 @@
+/-
+Copyright (c) 2025 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Mure
+-/
+
+/-!
+# A type system for NKI
+
+The polymorphic type system for NKI consisting of two sorts: `SNat` and `STyp`.
+
+`SNat` represents the natural numbers used to denote tensor shapes.
+`STyp` is the universe of types that includes tensors parametrized by `SNat`s.
+
+All type parameters are universally quantified and must be given up front.
+We then use De Bruijn indices in the form of `Fin` type to index type parameter binders.
+-/
+
+namespace KLR.NKI
+
+/--
+A type universe for denoting tensor shapes.
+
+`nnat` is the total number of universal quantifiers in scope.
+
+Example:
+`∀ n m n, n` ≡ `(SNat.param 0 : SNat 3)`
+-/
+inductive SNat (nnat : Nat)
+  | param (idx : Fin nnat)
+  | const (n : Nat)
+
+/--
+The universe of types.
+
+`nnat` is the total number of universal quantifiers in scope for `SNat`s.
+
+`ntyp` is the total number of universal quantifiers in scope for `STyp`s.
+
+Example:
+`∀ n m n, ∀ t1 t2, tensor[t1, 10, m]` ≡ `(STyp.tensor [SNat.const 10, SNat.param 1] (STyp.param 1) : STyp 3 2)`
+-/
+inductive STyp (nnat ntyp : Nat)
+  | param (idx : Fin nnat)
```

Comment:
Not sure about encoding the number of variables with dependent types. An external type checker seems easier for me to imagine working, but I look forward to seeing more examples. For example, it would be neat to typecheck something like

```
def transpose(x):
  return np.transpose(x, (3, 1))
```

This should accept a tensor of at least 4 dimensions, and type as something like
```
all n0 n1 . tensor[n0, n1, n2, n3, ...] -> tensor[n0, n3, n2, n1, ...]
```

How would this look with the dependent types above?

Reply:
You would need to quantify over `snat` and `styp` like so:

```lean
def List.swap (l : List α) (i₁ i₂ : Fin l.length) : List α :=
  (l.set i₁ l[i₁]).set i₂ l[i₂]

def transpose {nnat ntyp : Nat} {dtype : STyp nnat ntyp} {shapes : List (SNat nnat)}
  (hshapes : shapes.length > 3) : STyp nnat ntyp :=
  have h1 : 1 < shapes.length := by omega
  have h3 : 3 < shapes.length := by omega
  .func (.tensor shapes dtype) (.tensor (shapes.swap ⟨1, h1⟩ ⟨3, h3⟩ ) dtype)
```

However, you need to prove the length of `shapes` regardless in order to index into the `list`.

Reply:
I'd prefer to not do theorem proving when writing functions. We are trying to keep proofs and code entirely separate. This was suggested as a best practice by the CompCert folks and we've been trying to stick to that.

Reply:
Also, I should clarify that the need for less-than proofs in the transpose example has nothing to do with dependent De Bruijn indices but the need to index into lists. I could have defined it like this:
```lean
def List.swap (l : List α) (i₁ i₂ : Nat) : List α :=
  if h : i₁ < l.length ∧ i₂ < l.length then
    (l.set i₁ l[i₂]).set i₂ l[i₁]
  else
    l

def transpose {nnat ntyp : Nat} {dtype : Core.Dtype} {shapes : List (SNat nnat)} : STyp nnat ntyp :=
  .func (.tensor shapes dtype) (.tensor (List.swap shapes 1 3 ) dtype)
```